Server/JokesOnYou.Web.Api/Controllers/UserController.cs

```diff
@@ -1,48 +1,143 @@
 
-        public UserController(IUserService userService)
+        readonly IUserService _userService;
+        readonly ILogger<UserController> _logger;
+        readonly IMapper _mapper;   
+        readonly ITokenService _tokenService;
+        public UserController(IUserService userService, ILogger<UserController> logger, IMapper mapper, ITokenService tokenService)
         {
+            _logger = logger;
+            _mapper = mapper;
             _userService = userService;
+            _tokenService = tokenService;
         }
 
+        //ok so returning user is dumb, since we return sensitive data, we need better DTO system, Valve plz fix
         [HttpGet]
-        public async Task<ActionResult> GetUsers()
+        public ActionResult<IEnumerable<User>> GetUsers()
         {
-            throw new NotImplementedException();
+            return Ok(_userService.GetAll());
         }
 
+        //idk if its from body or Http get thing
         [HttpGet("{id}")]
-        public async Task<ActionResult> GetUser(string id)
+        public  ActionResult<User> GetUserById(string id)
         {
-            throw new NotImplementedException();
-        }
+            var user =  _userService.GetUserById(id);
 
-        [HttpPatch]
-        public async Task<ActionResult> UpdateUser(UserUpdateDTO userUpdateDTO)
+            if (user != null)
+            {
+                return user;
+            }
+            _logger.LogInformation($"User not found; id: {id}");
+            return NotFound();
+            
+        }
+        [HttpPut("{id}")]
+        public async Task<ActionResult> UpdateUser(string id, [FromBody]UserUpdateDTO userUpdateDTO)
         {
-            throw new NotImplementedException();
+            var user = _mapper.Map<User>(userUpdateDTO);
+
+
+            if (await _userService.UpdateUser(user))
+            {
+                return Ok();
+            }
+            _logger.LogInformation($"Something went wrong with updating user that had id {user.Id}");
+
+            return NotFound();
         }
 
+        [AllowAnonymous]
+        [HttpPost("authenticate")]
+        public async Task<ActionResult> Authenticate([FromBody] UserLoginDTO userDto)
+        {
+            User user;
+            if (string.IsNullOrWhiteSpace(userDto.Password))
+            {
+                return Unauthorized(); //or something else, idk
+            }
+
+            //maybe use a switch? should be much different
+            if (!string.IsNullOrWhiteSpace(userDto.Username))
+            {
+                user =await _userService.Authenticate(userDto.Username, userDto.Password);
+            }
+            else if (!string.IsNullOrWhiteSpace(userDto.Email))
+            {
+                user =await _userService.Authenticate(userDto.Email, userDto.Password);
+            }
+            else
+            {
+                return Unauthorized();
+            }
+
+            if (user == null)
```

Comment:
i think this will be handled inside the service

Reply:
So if i do the work in service will you accept my version of service? How will we work this out? I don't know how the other version service is going, should i do it all? should we wait for the other version to be done? Pong me on discord.